ring/partition_ring.go

```diff
@@ -75,11 +85,227 @@ func (r *PartitionRing) ActivePartitionForKey(key uint32) (int32, error) {
+	// We have to run the function anyway because the logic may filter out some INACTIVE partitions.
+	if size <= 0 || size >= len(r.desc.Partitions) {
+		size = len(r.desc.Partitions)
+	}
+
+	lookbackUntil := now.Add(-lookbackPeriod).Unix()
+
+	// Initialise the random generator used to select instances in the ring.
+	// There are no zones
+	random := rand.New(rand.NewSource(shardUtil.ShuffleShardSeed(identifier, "")))
+
+	// To select one more instance while guaranteeing the "consistency" property,
+	// we do pick a random value from the generator and resolve uniqueness collisions
+	// (if any) continuing walking the ring.
+	tokensCount := len(r.ringTokens)
+
+	result := make(map[int32]struct{}, size)
+	exclude := map[int32]struct{}{}
+
+	for len(result) < size {
+		start := searchToken(r.ringTokens, random.Uint32())
+		iterations := 0
+		found := false
+
+		for p := start; !found && iterations < tokensCount; p++ {
+			iterations++
+
+			// Wrap p around in the ring.
+			if p >= tokensCount {
+				p %= tokensCount
+			}
+
+			pid, ok := r.partitionByToken[Token(r.ringTokens[p])]
+			if !ok {
+				return nil, ErrInconsistentTokensInfo
+			}
+
+			// Ensure the partition has not already been included or excluded.
+			if _, ok := result[pid]; ok {
+				continue
+			}
+			if _, ok := exclude[pid]; ok {
+				continue
+			}
+
+			p, ok := r.desc.Partitions[pid]
+			if !ok {
+				return nil, ErrInconsistentTokensInfo
+			}
+
+			var (
+				shouldInclude bool
+				shouldExtend  bool
+			)
+
+			// Check whether the partition should be included in the result and/or the result set extended.
+			if lookbackPeriod > 0 {
+				if p.IsActive() {
+					// The partition is active. We should include it. The result set is extended only
+					// if the partition became active within the lookback window.
+					shouldInclude = true
+					shouldExtend = p.GetStateTimestamp() >= lookbackUntil
+				} else {
+					// The partition is inactive. We should include it (and then extend the result set)
+					// only if became inactive within the lookback window.
+					shouldInclude = p.GetStateTimestamp() >= lookbackUntil
+					shouldExtend = shouldInclude
+				}
+			} else {
+				// No lookback was provided. We only include active partitions and don't lookback.
+				shouldInclude = p.IsActive()
+				shouldExtend = false
+			}
+
+			// Either include or exclude the found partition.
+			if shouldInclude {
+				result[pid] = struct{}{}
+			} else {
+				exclude[pid] = struct{}{}
+			}
+
+			// Extend the shard, if requested.
+			if shouldExtend {
+				size++
+			}
+
+			// We can stop searching for other partitions only if this partition was included
+			// and no extension was requested, which means it's the "stop partition" for this cycle.
+			if shouldInclude && !shouldExtend {
+				found = true
+			}
+		}
+
+		// If we iterated over all tokens, and no new partition has been found, we can stop looking for more partitions.
```

Comment:
This means that shard is smaller than requested size. Should we return error in that case?

Reply:
I don't think so. The shard can be smaller than the requested size (e.g. size could be > than number of partitions in the ring). Am I missing anything?

Reply:
No, you're right.